// @shareai-lab/kode-sdk/dist/core/room.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Room = void 0;
class Room {
    constructor(pool) {
        this.pool = pool;
        this.members = new Map();
    }
    join(name, sessionId) {
        if (this.members.has(name)) {
            throw new Error(`Member already exists: ${name}`);
        }
        this.members.set(name, sessionId);
    }
    leave(name) {
        this.members.delete(name);
    }
    async say(from, text) {
        const mentions = this.extractMentions(text);
        if (mentions.length > 0) {
            // Directed message
            for (const mention of mentions) {
                const sessionId = this.members.get(mention);
                if (sessionId) {
                    const agent = this.pool.get(sessionId);
                    if (agent) {
                        await agent.send(`[from:${from}] ${text}`);
                    }
                }
            }
        }
        else {
            // Broadcast to all except sender
            for (const [name, sessionId] of this.members) {
                if (name !== from) {
                    const agent = this.pool.get(sessionId);
                    if (agent) {
                        await agent.send(`[from:${from}] ${text}`);
                    }
                }
            }
        }
    }
    getMembers() {
        return Array.from(this.members.entries()).map(([name, sessionId]) => ({ name, sessionId }));
    }
    extractMentions(text) {
        const regex = /@(\w+)/g;
        const mentions = [];
        let match;
        while ((match = regex.exec(text)) !== null) {
            mentions.push(match[1]);
        }
        return mentions;
    }
}
exports.Room = Room;
