// @shareai-lab/kode-sdk/dist/infra/provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnthropicProvider = void 0;
class AnthropicProvider {
    constructor(apiKey, model = 'claude-3-5-sonnet-20241022', baseUrl = 'https://api.anthropic.com') {
        this.apiKey = apiKey;
        this.model = model;
        this.baseUrl = baseUrl;
    }
    async complete(messages, opts) {
        const body = {
            model: this.model,
            messages: this.formatMessages(messages),
            max_tokens: opts?.maxTokens || 4096,
        };
        if (opts?.temperature !== undefined)
            body.temperature = opts.temperature;
        if (opts?.system)
            body.system = opts.system;
        if (opts?.tools && opts.tools.length > 0)
            body.tools = opts.tools;
        const response = await fetch(`${this.baseUrl}/v1/messages`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'x-api-key': this.apiKey,
                'anthropic-version': '2023-06-01',
            },
            body: JSON.stringify(body),
        });
        if (!response.ok) {
            const error = await response.text();
            throw new Error(`Anthropic API error: ${response.status} ${error}`);
        }
        const data = await response.json();
        return {
            role: 'assistant',
            content: data.content,
            usage: data.usage,
            stop_reason: data.stop_reason,
        };
    }
    async *stream(messages, opts) {
        const body = {
            model: this.model,
            messages: this.formatMessages(messages),
            max_tokens: opts?.maxTokens || 4096,
            stream: true,
        };
        if (opts?.temperature !== undefined)
            body.temperature = opts.temperature;
        if (opts?.system)
            body.system = opts.system;
        if (opts?.tools && opts.tools.length > 0)
            body.tools = opts.tools;
        const response = await fetch(`${this.baseUrl}/v1/messages`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'x-api-key': this.apiKey,
                'anthropic-version': '2023-06-01',
            },
            body: JSON.stringify(body),
        });
        if (!response.ok) {
            const error = await response.text();
            throw new Error(`Anthropic API error: ${response.status} ${error}`);
        }
        const reader = response.body?.getReader();
        if (!reader)
            throw new Error('No response body');
        const decoder = new TextDecoder();
        let buffer = '';
        while (true) {
            const { done, value } = await reader.read();
            if (done)
                break;
            buffer += decoder.decode(value, { stream: true });
            const lines = buffer.split('\n');
            buffer = lines.pop() || '';
            for (const line of lines) {
                if (!line.trim() || !line.startsWith('data: '))
                    continue;
                const data = line.slice(6);
                if (data === '[DONE]')
                    continue;
                try {
                    const event = JSON.parse(data);
                    if (event.type === 'content_block_start') {
                        yield { type: 'content_block_start', index: event.index, content_block: event.content_block };
                    }
                    else if (event.type === 'content_block_delta') {
                        yield { type: 'content_block_delta', index: event.index, delta: event.delta };
                    }
                    else if (event.type === 'content_block_stop') {
                        yield { type: 'content_block_stop', index: event.index };
                    }
                    else if (event.type === 'message_delta') {
                        yield { type: 'message_delta', delta: event.delta, usage: event.usage };
                    }
                    else if (event.type === 'message_stop') {
                        yield { type: 'message_stop' };
                    }
                }
                catch (e) {
                    // Skip invalid JSON
                }
            }
        }
    }
    formatMessages(messages) {
        return messages.map((msg) => ({
            role: msg.role === 'system' ? 'user' : msg.role,
            content: msg.content,
        }));
    }
}
exports.AnthropicProvider = AnthropicProvider;
